monkey/monkey_island/cc/repository/i_machine_repository.py

```diff
@@ -1,20 +1,74 @@
-from abc import ABC
-from typing import Optional, Sequence
+from abc import ABC, abstractmethod
+from ipaddress import IPv4Address
+from typing import Sequence
+
+from common.types import HardwareID
+from monkey_island.cc.models import Machine, MachineID
 
 
 class IMachineRepository(ABC):
-    # TODO define Machine object(ORM model)
-    def save_machine(self, machine: Machine):  # noqa: F821
-        pass
-
-    # TODO define Machine object(ORM model)
-    # TODO define or re-use machine state.
-    # TODO investigate where should the state be stored in edge or both edge and machine
-    def get_machines(
-        self,
-        id: Optional[str] = None,
-        ips: Optional[Sequence[str]] = None,
-        state: Optional[MachineState] = None,  # noqa: F821
-        is_island: Optional[bool] = None,  # noqa: F841
-    ) -> Sequence[Machine]:  # noqa: F821
-        pass
+    """A repository used to store and retrieve Machines"""
+
+    @abstractmethod
+    def create_machine(self) -> Machine:
```

Comment:
Most repositories use the term `save`, not `create`.

Reply:
I would agree with you but in this specific case we are not just saving the machine but we are returning the machine we are saving.

Reply:
Why does this take 0 parameters? Surely we know at least one of it's IP's

Reply:
As discussed, we replaced this with `get_next_id()` and changed `update_machine()` to `upsert_machine()`.